manpages/pulp_install.1
.TH PULP_INSTALL "1" "July 24" "" "User Commands"
.SH NAME
.B pull_install
Installation of PULPissimo project, an evolution of PULPino.
.SH SYNOPSIS         
.B pulp_install
-v [-c pulp|linux|linux32|linuxm]
[-p 0|1|2|3|4|5]
[-t y|n]
.SH DESCRIPTION
.TP 
.B SYSTEM REQUIREMENTS
In order to correctly install and use pulpissimo you should satisfy following requirements:
.RS 1.2i
.TP
.B  Ubuntu 16.04 or CentOS 7.
.TP
.B Mentor ModelSim tested with 10.6b
This is optional if you only use fpga target
.TP
.B python3.4
.TP
.B Vivado 19
.RE          
.PP
.B pulp_install
is a bash script that install and configure 
.B PULPissimo project 
in the directory in which it
is runned, particularly it download and configure following repository:
.PP                    
.TP
.B pulp-riscv-gnu-toolchain installation 
This toolchain is available at <https://github.com/pulp-platform/pulp-riscv-gnu-toolchain-pulp-riscv-gnu-toolchain> It provides the
.B RISCV C and C++ cross-compiler.
After the installation of all the necessary Ubuntu (or CentOS) 
.B dependencies
, the script clones the git repository of the toolchain and configures the correct cross-compiler according to the option -c|--cross-compiler set as command line argument. With the -t|--test-suite command line argument, instead, the user can decide if install or not the De-jaGnu test suite for RISC-V, that is a front-end to write tests for any pro-gram with a portable interface.  

In this step, the environment variables PULP_RISCV_GCC_TOOLCHAIN and  VSIM_PATH are set. The installation directory of the riscv toolchain is "/opt/riscv/", so this is the value assigned to PULP_RISCV_GCC_TOOLCHAIN variable. Instead, VSIM_PATH is set to "(pulpissimo_root)/pulpissimo". Obviously, also other minor variables are set in this step. All these env variables are also added to ~/.bash_profile file so as to make them permanent every time a new bash shell is create.

.TP 
.B PULP SDK installation
The PULP SDK is available ati <https://github.com/pulp-platform/pulp-sdk.git> and after the installation of all the dependencies (such as specific version of Python) the repository is cloned.  In this stage, target and platform (rtl simulator, fpga, virtual platform) are chosen, so that the SDK can be built properly. This is a very important step, because the SDK allows to compile and run applications. To do this, it needs to be initialized by sourcing the "sourceme.sh" file which is inside the installation folder. Also in this stage, other env variables are set and saved permanently.

.TP 
.B PULP Builder installation
The PULP Builder repository is available at <https://github.com/pulp-platform/pulp-builder.git> and is used to build application for pulp project.  After the clone of the repository the builder is configured.

.TP
.B PULPissimo installation 
PULPissimo project repository is available at <https://github.com/pulp-platform/pulpissimo.git>. This is the most important step, since the PULPissimo project contains all the files required to build the rtl simulator or to map the system on fpga. The repository also contains all the files of IPs which compose the PULP system, including the RI5CY core files, and documentation, fpga support, testbench, simulation and examples.
Files related to the ZCU102 will be used later in another script aimed to facilitate the fpga configuration process and the generation of the bitstream to flash into the board.

.TP
.B PULPissimo example
Example of application project can be found at <https://github.com/pulp-platform/pulp-rt-examples.git>. The script downloads this repository and build hello application by default (it is possible to change the application to build). These example applications will be used later in another script for the fpga.
.TP
.B Virtual platform installation 
This is the last step that has been added for completeness. This step prepares the PULP environment for the virtual platform.


.SH OPTION
A list of flags and their descriptions:

.TP
.B v|--verbose
verbose option, script print many others information
.TP
.B c|--cross_compiler [pulp|linux|linux32|linuxm]
This option define what cross compiler the script install:
.RS 1.2i
.TP
.B pulp
newlib cross-compiler for all pulp variants
This will use the multilib support to build the libraries for 
the various cores (riscy, zeroriscy and so on). The right libraries 
will be selected depending on which compiler options you use.
.TP
.B newlib 
Newlib cross-compiler, You should now be able to use riscv-gcc and its cousins.
.TP
.B linux
Linux cross compiler 64 bit
Supported architectures are rv64i plus standard extensions (a)tomics, 
(m)ultiplication and division, (f)loat, (d)ouble, or (g)eneral for MAFD.
Supported ABIs are ilp32 (32-bit soft-float), ilp32d (32-bit hard-float), ilp32f 
(32-bit with single-precision in registers and double in memory, niche use only), 
lp64 lp64f lp64d (same but with 64-bit long and pointers).
.TP 
.B linux32 
Linux cross compiler 32 bit
Supported architectures are rv32i plus standard extensions (a)tomics, 
(m)ultiplication and division, (f)loat, (d)ouble, or (g)eneral for MAFD.
Supported ABIs are ilp32 (32-bit soft-float), ilp32d (32-bit hard-float), ilp32f 
(32-bit with single-precision in registers and double in memory, niche use only), 
lp64 lp64f lp64d (same but with 64-bit long and pointers).
.TP
.B linuxm 
Linux cross-compiler, both 32 and 64 supported
.RE

.TP 
.B -p|--part_install [0|1|2|3|4|5]
Default is set to 0. Set this argument from 0 to 3 to decide the starting point of the installation:
.RS 1.2i
.PP
.B 0
start from scratch
.PP
.B 1
start after the toolchain
.PP
.B 2
start after the sdk
.PP
.B 3
start after pulp-builder
.PP
.B 4
start after test building (hello)
.PP
.B 5
only virtual platform
.RE

.TP 
.B -t|--test_suite [y|n] 
Decide if install test suite or not, this test suite 

.PP
.SH ENVIRONMENT      
All setted variable are added to .bash_profile file that is loaded each bash restart.
.SH SEE ALSO
.BR pulp_app 1 ,
.BR pulp_script 1
.SH BUGS            
.SH CREDITS 
Created by Elia Ribaldone, Luca Fiore and Marcello Neri


pulp_script/pulp_install.sh
# This file install pulpissimo and all dependecies:
# 
# pulpissimo dependencies:
#		installation of pulp-builder:
#			https://github.com/pulp-platform/pulp-builder/blob/master/README.md
# 			pulp-builder dependencies:
#  				installation of pulp-riscv-gnu-toolchain 
#					https://github.com/pulp-platform/pulp-riscv-gnu-toolchain
# rtl requirements:
# 		Ubuntu 16.04 and CentOS 7.
#		Mentor ModelSim tested with 10.6b
#		python3.4  (and pyyaml)
#		installation of SDK (software development kit):
#			https://github.com/pulp-platform/pulp-sdk/blob/master/README.md
#		
source $BASHLIB/ccommon.sh

gitclone() {
	# This function is used to avoid recloning of reposistory each time
	# that the installation is redone due to some error
	# $1 comando
	# $2 clone dir
	# $3 log file
	cmd=$1
	clone_name=$2
	log_file=$3

	if ! test -d $CLONE/$clone_name; then
		Print_verbose "[*] 		Download of $clone_name" $verbose	
		cd $CLONE
		#mon_run "$cmd" "$LOG_DIR/log/$log_file.txt" 1 $LINENO
		$cmd
		cd ..
	fi
	if test -d $clone_name; then
		sudo rm -rf $clone_name
		
	fi
	cp -r $CLONE/$clone_name .
}

UsageExit() {
	   echo \
"pulp_install help:\n \
	Options:
	   	-v|--verbose   
			verbose option, script print many others information

		-c|--cross_compiler [pulp|linux|linux32|linuxm]
			This option define what cross compiler the script install:
				pulp ->  newlib cross-compiler for all pulp variants
					 This will use the multilib support to build the libraries for 
					 the various cores (riscy, zeroriscy and so on). The right libraries 
					 will be selected depending on which compiler options you use.
				newlib -> Newlib cross-compiler, You should now be able
					 to use riscv-gcc and its cousins.
				linux -> Linux cross compiler 64 bit
					 Supported architectures are rv64i plus standard extensions (a)tomics, 
					 (m)ultiplication and division, (f)loat, (d)ouble, or (g)eneral for MAFD.
					 Supported ABIs are ilp32 (32-bit soft-float), ilp32d (32-bit hard-float), ilp32f 
					 (32-bit with single-precision in registers and double in memory, niche use only), 
					 lp64 lp64f lp64d (same but with 64-bit long and pointers).
				linux32 -> Linux cross compiler 32 bit
					 Supported architectures are rv32i plus standard extensions (a)tomics, 
					 (m)ultiplication and division, (f)loat, (d)ouble, or (g)eneral for MAFD.
					 Supported ABIs are ilp32 (32-bit soft-float), ilp32d (32-bit hard-float), ilp32f 
					 (32-bit with single-precision in registers and double in memory, niche use only), 
					 lp64 lp64f lp64d (same but with 64-bit long and pointers).
				linuxm ->  Linux cross-compiler, both 32 and 64 supported
		-p|--part_install [0|1|2|3|4|5]
			Default is set to 0. Set this argument from 0 to 5 to decide the starting point of the installation:
				0 -> start from scratch
				1 -> start after the toolchain
				2 -> start after the sdk
				3 -> start after pulp-builder
				4 -> test (hello)
				5 -> virtual platform
		-t|--test_suite [y|n]
			Decide if install test suite or not, this test suite 

"
	   exit 1;
}



########### Variable
SYSTEM="ubuntu"
DIR="$(dirname "$(readlink -f "${BASH_SOURCE[0]}")")"
INSTALL_DIR="/opt/riscv"
LOG_DIR=$DIR
PART=0
# cross compiler variables
CROSS_COMPILER="pulp" # decide the kind of compiler to be installed
TEST_SUITE=1 # if 1 test suite is installed
TEST_SUITE_BIT=64 
# pulpussimo
PULPISSIMO_ROOT="$DIR/pulpissimo"
CLONE="clone_dir"
########### end variable

#export_var "PULP_RISCV_GCC_TOOLCHAIN" "$INSTALL_DIR"
#export_var "VSIM_PATH" "$PULPISSIMO_ROOT"
export PULP_RISCV_GCC_TOOLCHAIN=$INSTALL_DIR
export VSIM_PATH=$PULPISSIMO_ROOT
export PATH=$PATH:$INSTALL_DIR/bin
#echo "export PATH=$PATH:$INSTALL_DIR/bin" >> ~/.bash_profile

########## create error log file ###########
rm -rf error_monitor.txt error_log.txt trace_command.txt
touch error_monitor.txt error_log.txt trace_command.txt

########## ARGUMENT PARSING #######################################

TEMP=`getopt -o hp:vc:t: --long help,part:,verbose,cross_compiler:,test_suite: -- "$@"`
eval set -- "$TEMP"
echo $@

while true; do
	case $1 in
		-p|--part)
			# select the part from which to start the installation	
			shift
			PART=$1
			if [[ $1 -lt 0 ]] || [[ $1 -gt 3 ]]; then
				UsageExit;
			fi
			shift
			;;
		-v|--verbose)
			# verbose print using print_verbose
			verbose=1; shift
			;;
		-c|--cross_compiler)
			# decide cross compiler for gnu toolchain
			shift 
			case $1 in 
				pulp)
					# newlib cross-compiler for all pulp variants
					# This will use the multilib support to build the libraries for 
					# the various cores (riscy, zeroriscy and so on). The right libraries 
					# will be selected depending on which compiler options you use.
					CROSS_COMPILER="pulp";;
				newlib)
					# Newlib cross-compiler, You should now be able to use riscv-gcc and its cousins.
					CROSS_COMPILER="newlib";;
				linux|linux64)
					# Linux cross-compiler
					# Supported architectures are rv32i or rv64i plus standard extensions (a)tomics, 
					# (m)ultiplication and division, (f)loat, (d)ouble, or (g)eneral for MAFD.
					# Supported ABIs are ilp32 (32-bit soft-float), ilp32d (32-bit hard-float), ilp32f 
					# (32-bit with single-precision in registers and double in memory, niche use only), 
					# lp64 lp64f lp64d (same but with 64-bit long and pointers).
                    CROSS_COMPILER="linux64";;
				linux32)
					# as before but 32bit
					CROSS_COMPILER="linux32"
					;;
				linuxm)
					# Linux cross-compiler, both 32 and 64 supported
					CROSS_COMPILER="linuxm"
					;;
				*)
					echo "[!!] error on cross compiler option!!"
					UsageExit;
					;;
			esac
			shift
			;;
		-t|--test_suite)
			# The DejaGnu test suite has been ported to RISC-V. This can run with GDB simulator for elf
			# toolchain or Qemu for linux toolchain, and GDB simulator doesn't support
			# floating-point.
			shift 
			case $1 in
				y)
					TEST_SUITE=1;;
				n) 
					TEST_SUITE=0;;
				y32)
					TEST_SUITE=1;
					TEST_SUITE_BIT=32;;
				*)
					echo "[!!] error on test suite option, only y/n are allowed!!"
					UsageExit;;
			esac
			shift
			;;
		-h|--help)
			UsageExit
			;;
		--)
			break;;
		*)
			UsageExit
			;;
	esac
done
############# enlarge the window to fit any writings
mon_run "sudo apt-get install xterm -y" $LOG_DIR/log/xterm.txt  1 $LINENO 
resize -s 30 145

############# Check operating system
Print_verbose "[*] Operative system check:" $verbose
if [[ $(uname -a | grep -i ubuntu | wc -l) = 1 ]]; then
	SYSTEM=ubuntu
elif [[ $(uname -a | grep -i centos | wc -l) = 1 ]]; then
	SYSTEM=centos
else
	echo "System not supported!!!"
	exit 1;
fi
Print_verbose "[*]		Your os is $SYSTEM" $verbose
############# Git instllation

if ! hash git 2>/dev/null ; then
	Print_verbose "[*] Install git"
	mon_run "sudo apt-get install git -y" $LOG_DIR/log/git.txt 1 $LINENO 
	# when git install finish this command finished
fi

############# installation of pulp-riscv-gnu-toolchain
# da https://github.com/pulp-platform/pulp-riscv-gnu-toolchain
# RISC-V GNU Compiler Toolchain supports two build modes:
#		generic ELF/Newlib toolchain
#       more sophisticated Linux-ELF/glibc toolchain

# Dipendecies of pulp-riscv-gnu-toolchain
if [[ $PART -eq 0 ]]; then
	Print_verbose "[+] Installation of pulp-riscv-gnu-toolchain: " $verbose
	Print_verbose "[*] 		Install dipendecies" $verbose
	if [[ $SYSTEM = "ubuntu" ]]; then
		mon_run "sudo apt-get install autoconf automake autotools-dev curl\
			libmpc-dev libmpfr-dev libgmp-dev gawk build-essential\
			 bison flex texinfo gperf libtool patchutils bc zlib1g-dev -y" $LOG_DIR/log/dep_toolchain.txt 1 $LINENO
	else
		mon_run "sudo yum install autoconf automake libmpc-devel mpfr-devel \
			gmp-devel gawk  bison flex texinfo patchutils \
			gcc gcc-c++ zlib-devel -y" $LOG_DIR/log/dep_toolchain.txt 1 $LINENO
	fi
	mkdir -p $CLONE
	# Download of toolchain
	gitclone "git clone --recursive https://github.com/pulp-platform/pulp-riscv-gnu-toolchain" "pulp-riscv-gnu-toolchain" "toolchain"

	sudo mkdir -p $INSTALL_DIR
	cd pulp-riscv-gnu-toolchain
	export PATH=$PATH:$INSTALL_DIR/bin
	echo "export PATH=$PATH:$INSTALL_DIR/bin" >> ~/.bash_profile
	# Install cross compiler 

	Print_verbose "[*] 		Install selected cross compiler: $CROSS_COMPILER" $verbose
	case $CROSS_COMPILER in
		pulp)
			mon_run "sudo ./configure --prefix=$INSTALL_DIR --with-arch=rv32imc --with-cmodel=medlow --enable-multilib"\
			       	$LOG_DIR/log/cross_compiler.txt 1 $LINENO
			mon_run "sudo make" $LOG_DIR/log/cross_compiler.txt 0 $LINENO
			;;
		newlib)
			mon_run "sudo ./configure --prefix=$INSTALL_DIR"  $LOG_DIR/log/cross_compiler.txt 1 $LINENO
			mon_run "sudo make" $LOG_DIR/log/cross_compiler.txt 0 $LINENO
			;;
		linux64)
			mon_run "sudo ./configure --prefix=$INSTALL_DIR" $LOG_DIR/log/cross_compiler.txt 1  $LINENO
			mon_run "make linux" $LOG_DIR/log/cross_compiler.txt 0 $LINENO
			;;
		linux32)
			mon_run "./configure --prefix=$INSTALL_DIR --with-arch=rv32g --with-abi=ilp32d" $LOG_DIR/log/cross_compiler.txt 1 $LINENO
			mon_run "make linux" $LOG_DIR/log/cross_compiler.txt 0 $LINENO
			;;
		linuxm)
			mon_run "./configure --prefix=$INSTALL_DIR --enable-multilib" $LOG_DIR/log/cross_compiler.txt 1 $LINENO
			mon_run "make linux" $LOG_DIR/log/cross_compiler.txt 0 $LINENO
			;;
		*)
			echo "[!!] Inter error on cross_compiler options"
			UsageExit
			;;
	esac


	# Install Test Suite

	if [[ $TEST_SUITE -eq 1 ]]; then
		Print_verbose "[*] 		Install Test Suite\n[*]	Configure" $verbose
	
		case $CROSS_COMPILER in
			pulp|newlib)
				if [[ $TEST_SUITE_BIT -eq 64 ]]; then
					mon_run "sudo ./configure --prefix=$RISCV --disable-linux --with-arch=rv64ima"\
					       	$LOG_DIR/log/test_suite.txt 1 $LINENO
				else
					mon_run "sudo ./configure --prefix=$RISCV --disable-linux --with-arch=rv32ima"\
					       	$LOG_DIR/log/test_suite.txt 1 $LINENO
				fi
				Print_verbose "[*] make" $verbose
			
				mon_run "sudo make newlib" $LOG_DIR/log/test_suite.txt 0 $LINENO

				mon_run "sudo make check-gcc-newlib" $LOG_DIR/log/test_suite.txt 0 $LINENO
				;;
			linux|linux32|linux64)
				mon_run "./configure --prefix=$RISCV" $LOG_DIR/log/test_suite.txt 1 $LINENO
			 
				mon_run "make linux" $LOG_DIR/log/test_suite.txt 0 $LINENO

				mon_run "make check-gcc-linux" $LOG_DIR/log/test_suite.txt 0 $LINENO
				;;
		esac
	fi

	export_var "PULP_RISCV_GCC_TOOLCHAIN" "$INSTALL_DIR"
	export_var "VSIM_PATH" "$PULPISSIMO_ROOT"

	cd ../ # exit from tooolchain
fi # end of toolchain, PART=0

############# PULP SDK build process
## see https://github.com/pulp-platform/pulp-sdk/blob/master/README.md

if [[ $PART -le 1 ]]; then
	Print_verbose "[+] Installaition of SDK build process " $verbose

	###### Install dipendecies for SDK build process
	Print_verbose "[*] 		Install dipendecies for SDK build process " $verbose
	mon_run "sudo apt install git python3-pip python-pip gawk texinfo \
		libgmp-dev libmpfr-dev libmpc-dev swig3.0 libjpeg-dev lsb-core \
		doxygen python-sphinx sox graphicsmagick-libmagick-dev-compat \
		libsdl2-dev libswitch-perl libftdi1-dev cmake scons libsndfile1-dev -y"\
		$LOG_DIR/log/sdk_dep.txt 1 $LINENO
	mon_run "sudo pip3 install openpyxl" $LOG_DIR/log/sdk_dep.txt 0 $LINENO
	mon_run "sudo pip3 install artifactory twisted prettytable sqlalchemy \
		pyelftools openpyxl==2.6.4 xlsxwriter pyyaml numpy configparser pyvcd"\
		$LOG_DIR/log/sdk_dep.txt 0 $LINENO
	mon_run "sudo pip2 install configparser" $LOG_DIR/log/sdk_dep.txt 0 $LINENO

	######  Installation of gcc5 and g++5
	Print_verbose "[*]  	Installation of gcc5 and g++5" $verbose
	mon_run "sudo apt-get install gcc-5 g++-5 -y" $LOG_DIR/log/gcc++.txt 1 $LINENO
	mon_run "sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-5 10" $LOG_DIR/log/gcc++.txt 0 $LINENO
	mon_run "sudo update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-5 10" $LOG_DIR/log/gcc++.txt 0 $LINENO
	mon_run "sudo update-alternatives --install /usr/bin/cc cc /usr/bin/gcc 30" $LOG_DIR/log/gcc++.txt 0 $LINENO
	mon_run "sudo update-alternatives --set cc /usr/bin/gcc" $LOG_DIR/log/gcc++.txt 0 $LINENO
	mon_run "sudo update-alternatives --install /usr/bin/c++ c++ /usr/bin/g++ 30" $LOG_DIR/log/gcc++.txt 0 $LINENO
	mon_run "sudo update-alternatives --set c++ /usr/bin/g++" $LOG_DIR/log/gcc++.txt 0 $LINENO

	gitclone "git clone https://github.com/pulp-platform/pulp-sdk.git -b master --progress" "pulp-sdk"  "sdk_downl"

	cd pulp-sdk
	Print_verbose "[*] 		Target and platform selection" $verbose
	source configs/pulpissimo.sh
	# I have to choose the platform: board,fpga,gvsoc,hsas,rtl
	source configs/platform-rtl.sh
	Print_verbose "[*] 		Build" $verbose
	mon_run "make all"  $LOG_DIR/log/sdk_build.txt 1 $LINENO

	cd ../ #exit from sdk
	
fi #end of sdk, PART<=1

############# Pulp builder install
if [[ $PART -le 2 ]]; then

	gitclone "git clone https://github.com/pulp-platform/pulp-builder.git --progress" "pulp-builder" "pulp-builder"

	cd pulp-builder
	mon_run "git checkout 0e51ae60d66f4ec326582d63a9fcd40ed2a70e15" $LOG_DIR/log/pulp_builder.txt 1 $LINENO
	mon_run "source configs/pulpissimo.sh" $LOG_DIR/log/pulp_builder.txt 0 $LINENO
	mon_run "./scripts/clean" $LOG_DIR/log/pulp_builder.txt 0 $LINENO
	mon_run "./scripts/update-runtime" $LOG_DIR/log/pulp_builder.txt 0 $LINENO
	mon_run "./scripts/build-runtime" $LOG_DIR/log/pulp_builder.txt 0 $LINENO
	mon_run "source sdk-setup.sh" $LOG_DIR/log/pulp_builder.txt 0 $LINENO
	mon_run "source configs/rtl.sh" $LOG_DIR/log/pulp_builder.txt 0 $LINENO
	cd ..
fi #end of pulp-builder, PART<=2

########### Pulpissimo 
if [[ $PART -le 3 ]]; then
	Print_verbose "[*] 	Installation of Pulpissimo"
	gitclone "git clone https://github.com/pulp-platform/pulpissimo.git --progress " "pulpissimo" "pulpissimo"
	cd pulpissimo
	mon_run "./update-ips" $LOG_DIR/log/update_ips.txt 1 $LINENO

	##### MODELSIM ####
	source setup/vsim.sh
	mon_run "make clean build" $LOG_DIR/log/vsim_clean_build.txt 1 $LINENO
	cd ..
fi #end of pulpissimo, PART<=3


######## test
if [[ $PART -le 4 ]]; then
	gitclone "git clone https://github.com/pulp-platform/pulp-rt-examples.git --progress " "pulp-rt-examples" "pulp-rt-examples"
	cd pulp-rt-examples/hello
	mon_run "make clean all run" $LOG_DIR/log/make_hello.txt 1 $LINENO
	cd ..
fi #end of test


############ virtual platform
if [[ $PART -le 5 ]]; then
	cd ./pulp-builder
	git checkout 7bd925324fcecae2aad9875f4da45b27d8356796
	source configs/pulpissimo.sh
	./scripts/build-gvsoc
	source sdk-setup.sh
	source configs/gvsoc.sh
	cd ..
	#mon_run "make conf" $LOG_DIR/log/make_conf.txt 1 $LINENO
fi #end of virtual platform




